### ISSUE TITLE
Crash by java.util.regex.PatternSyntaxException

### ISSUE THREAD

**Issue by Hirschen** [kefirfromperm/kefirbb#45]
I tried to add KefirBB to my android project and with the default initializer I got this exception:

```
FATAL EXCEPTION: main
 Process: se.avanzabank.androidapplikation.debug, PID: 488
 java.lang.ExceptionInInitializerError
     at org.kefirsf.bb.PatternElementFactory.create(PatternElementFactory.java:43)
     at org.kefirsf.bb.ProcessorBuilder.createPattern(ProcessorBuilder.java:139)
     at org.kefirsf.bb.ProcessorBuilder.createCode(ProcessorBuilder.java:97)
     at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:67)
     at org.kefirsf.bb.ProcessorBuilder.build(ProcessorBuilder.java:33)
     at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:52)
     at org.kefirsf.bb.BBProcessorFactory.createFromResource(BBProcessorFactory.java:63)
     at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.populateWithArticle(PlaceraArticleFragment.java:226)
     at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.lambda$null$235(PlaceraArticleFragment.java:143)
     at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment$$Lambda$4.run(Unknown Source)
     at android.os.Handler.handleCallback(Handler.java:751)
     at android.os.Handler.dispatchMessage(Handler.java:95)
     at android.os.Looper.loop(Looper.java:154)
     at android.app.ActivityThread.main(ActivityThread.java:6077)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
  Caused by: java.util.regex.PatternSyntaxException: In a character range [x-y], x is greater than y near index 8
 #([\w&-=]|(%\p{XDigit}{2}))*
         ^
     at java.util.regex.Pattern.compileImpl(Native Method)
     at java.util.regex.Pattern.compile(Pattern.java:1340)
     at java.util.regex.Pattern.<init>(Pattern.java:1324)
     at java.util.regex.Pattern.compile(Pattern.java:946)
     at org.kefirsf.bb.proc.ProcUrl.<clinit>(ProcUrl.java:21)
     	... 17 more
```

Any ideas how to fix this?

**Comment by kefirfromperm** [comment]
Please try it. If is it OK I'll release it this evening. Could you say how I can test KefirBB on Android? How can I run unit tests?
[kefirbb-1.4-SNAPSHOT.zip](https://github.com/kefirfromperm/kefirbb/files/426642/kefirbb-1.4-SNAPSHOT.zip)

**Comment by Hirschen** [comment]
Thank you for amazing response time 😄 

There is a doc site for that:
https://developer.android.com/training/testing/unit-testing/local-unit-tests.html

If you're using gradle I think it would just be easy to do a small subproject that applies the android plugins and follow that guide. I don't think you need to do any real android programming for this but if gradle complains you can just generate an empty android project with Android Studio.

I'll test the snapshot later today and come back!

**Comment by Hirschen** [comment]
I tried the snapshot but it contains logging frameworks and the such that is missing in Android so I get lots of different `ClassNotFoundException`. At first it was slf4j but adding that lead to missing `ch.qos.logback.classic.LoggerContext`

**Comment by kefirfromperm** [comment]
I don't think this guide is useful. Because the problem is given by differences between Oracle JVM and Android JVM.

**Comment by Hirschen** [comment]
Ah yes, that is true. However for you to run on the Android JVM you will need to launch an emulator to run the code basically and that is a bit less useful than a unit test...

**Comment by kefirfromperm** [comment]
I don't know why it require `ch.qos.logback.classic.LoggerContext`. The logback dependencies is needed only for the test environment and doesn't use in main code. Could you attach a stacktrace?

**Comment by Hirschen** [comment]
```
08-19 13:11:44.338 E/AndroidRuntime: FATAL EXCEPTION: main
                                     Process: se.avanzabank.androidapplikation.debug, PID: 25399
                                     java.lang.NoClassDefFoundError: Failed resolution of: Lch/qos/logback/core/ContextBase;
                                         at org.slf4j.impl.StaticLoggerBinder.getSingleton(StaticLoggerBinder.java:68)
                                         at org.slf4j.LoggerFactory.getSingleton(LoggerFactory.java:189)
                                         at org.slf4j.LoggerFactory.bind(LoggerFactory.java:112)
                                         at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:105)
                                         at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:235)
                                         at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:208)
                                         at org.kefirsf.bb.proc.ProcPattern.<init>(ProcPattern.java:15)
                                         at org.kefirsf.bb.ProcessorBuilder.createPattern(ProcessorBuilder.java:141)
                                         at org.kefirsf.bb.ProcessorBuilder.createCode(ProcessorBuilder.java:97)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:67)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.build(ProcessorBuilder.java:33)
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:52)
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:42)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.populateWithArticle(PlaceraArticleFragment.java:224)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.lambda$null$235(PlaceraArticleFragment.java:141)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment$$Lambda$4.run(Unknown Source)
                                         at android.os.Handler.handleCallback(Handler.java:751)
                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                         at android.os.Looper.loop(Looper.java:154)
                                         at android.app.ActivityThread.main(ActivityThread.java:6077)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                                      Caused by: java.lang.ClassNotFoundException: Didn't find class "ch.qos.logback.core.ContextBase" on path: DexPathList[[zip file "/data/app/se.avanzabank.androidapplikation.debug-2/base.apk"],nativeLibraryDirectories=[/data/app/se.avanzabank.androidapplikation.debug-2/lib/arm64, /system/lib64, /vendor/lib64]]
                                         at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                         at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
                                         at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
                                         at org.slf4j.impl.StaticLoggerBinder.getSingleton(StaticLoggerBinder.java:68) 
                                         at org.slf4j.LoggerFactory.getSingleton(LoggerFactory.java:189) 
                                         at org.slf4j.LoggerFactory.bind(LoggerFactory.java:112) 
                                         at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:105) 
                                         at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:235) 
                                         at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:208) 
                                         at org.kefirsf.bb.proc.ProcPattern.<init>(ProcPattern.java:15) 
                                         at org.kefirsf.bb.ProcessorBuilder.createPattern(ProcessorBuilder.java:141) 
                                         at org.kefirsf.bb.ProcessorBuilder.createCode(ProcessorBuilder.java:97) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:67) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.build(ProcessorBuilder.java:33) 
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:52) 
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:42) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.populateWithArticle(PlaceraArticleFragment.java:224) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.lambda$null$235(PlaceraArticleFragment.java:141) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment$$Lambda$4.run(Unknown Source) 
                                         at android.os.Handler.handleCallback(Handler.java:751) 
                                         at android.os.Handler.dispatchMessage(Handler.java:95) 
                                         at android.os.Looper.loop(Looper.java:154) 
                                         at android.app.ActivityThread.main(ActivityThread.java:6077) 
                                         at java.lang.reflect.Method.invoke(Native Method)
```

**Comment by kefirfromperm** [comment]
Which version of slf4j-api did you add to dependencies?
I use
```xml
<dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.13</version>
</dependency>
```

**Comment by Hirschen** [comment]
No sure actually because I just grabbed a jar from somewhere. Added 1.7.21 as gradle dependency and got this instead:
```
08-19 13:57:19.338 E/AndroidRuntime: FATAL EXCEPTION: main
                                     Process: se.avanzabank.androidapplikation.debug, PID: 23351
                                     java.lang.NoClassDefFoundError: Failed resolution of: Lorg/slf4j/LoggerFactory;
                                         at org.kefirsf.bb.proc.ProcPattern.<init>(ProcPattern.java:15)
                                         at org.kefirsf.bb.ProcessorBuilder.createPattern(ProcessorBuilder.java:141)
                                         at org.kefirsf.bb.ProcessorBuilder.createCode(ProcessorBuilder.java:97)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:67)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63)
                                         at org.kefirsf.bb.ProcessorBuilder.build(ProcessorBuilder.java:33)
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:52)
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:42)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.populateWithArticle(PlaceraArticleFragment.java:224)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.lambda$null$235(PlaceraArticleFragment.java:141)
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment$$Lambda$4.run(Unknown Source)
                                         at android.os.Handler.handleCallback(Handler.java:751)
                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                         at android.os.Looper.loop(Looper.java:154)
                                         at android.app.ActivityThread.main(ActivityThread.java:6077)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                                      Caused by: java.lang.ClassNotFoundException: Didn't find class "org.slf4j.LoggerFactory" on path: DexPathList[[zip file "/data/app/se.avanzabank.androidapplikation.debug-1/base.apk"],nativeLibraryDirectories=[/data/app/se.avanzabank.androidapplikation.debug-1/lib/arm64, /system/lib64, /vendor/lib64]]
                                         at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                         at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
                                         at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
                                         at org.kefirsf.bb.proc.ProcPattern.<init>(ProcPattern.java:15) 
                                         at org.kefirsf.bb.ProcessorBuilder.createPattern(ProcessorBuilder.java:141) 
                                         at org.kefirsf.bb.ProcessorBuilder.createCode(ProcessorBuilder.java:97) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:67) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.createScope(ProcessorBuilder.java:63) 
                                         at org.kefirsf.bb.ProcessorBuilder.build(ProcessorBuilder.java:33) 
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:52) 
                                         at org.kefirsf.bb.BBProcessorFactory.create(BBProcessorFactory.java:42) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.populateWithArticle(PlaceraArticleFragment.java:224) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment.lambda$null$235(PlaceraArticleFragment.java:141) 
                                         at se.avanzabank.androidapplikation.news.fragment.PlaceraArticleFragment$$Lambda$4.run(Unknown Source) 
                                         at android.os.Handler.handleCallback(Handler.java:751) 
                                         at android.os.Handler.dispatchMessage(Handler.java:95) 
                                         at android.os.Looper.loop(Looper.java:154) 
                                         at android.app.ActivityThread.main(ActivityThread.java:6077) 
                                         at java.lang.reflect.Method.invoke(Native Method) 
```

**Comment by Hirschen** [comment]
Sorry wrong dependency. By mistake I used site instead of api module. Now it runs with the snapshot!

**Comment by Hirschen** [comment]
I'm doing regular work at the same time as this. I'm not too good at multitasking 😄

**Comment by kefirfromperm** [comment]
I've just released version 1.4

**Comment by Hirschen** [comment]
Great! Thanks :)
